(TK-209) Add :is_running to status service output

rlinehan opened the pull request:
This PR makes the following changes:
- unwrap exceptions from pmap so that they can be handled by our middleware
- add an `:is_running` key to the status service output
- add a `:detail_level` key to the status service output
- add a schema for the `StatusCallbackResponse`
- change hyphens to underscores in JSON output

rlinehan committed:
(TK-209) Add :is-running to status service output

This commit adds an `:is-running` key to the status service output. It gets
the value for this key from the output of the status callback function that a
service registers. If the callback function does not return a map with an
:is-running key, or if the value for this key is not a boolean, the status
service returns :unknown.

This commit also adds a schema for the status callback response. The status
service does not actually validate that the result of calling the callback
function for a service matches this schema, but this schema can be used by a
service registering a callback function to validate the functions output.

rlinehan committed:
(maint) Improve error response

Exceptions for pmaps get wrapped in `java.util.concurrent.ExecutionException`
since the exception happens in a future. However, this meant that if a schema
error happened when calling one of the status callback functions, it would get
wrapped by this and our `wrap-schema-error` middleware wouldn't catch it. This
commit unwraps the `ExecutionException` and rethrows the original exception.
It also pulls some additional keys out of the schema exception so that we get
a more useful error message in the case of an unexpected schema error.

rlinehan committed:
(TK-209) Add :detail-level to status service response

Also update docstring.

rlinehan committed:
(maint) Change hyphens to underscores in json output

We've typically standardized on having keys in maps in json use underscores
rather than hyphens. This commit changes our code to do this.

cprice404 commented:
LGTM, need to add a test to describe the behavior when multiple services are registered and one of them misbehaves, and change the implementation accordingly.

Hmmm... I just noticed that we have a separate ticket for that: TK-175.  Maybe it's OK to punt that work until that ticket comes in, then.

If that's the way we're planning to go then maybe for this one all that's left is to consider consolidating the callback response schema with the service status schema.  Changing that later would be a breaking API change, so it'd be good to make a call on it before we send out our announcement.  (The decision will also affect the documentation.)

Looking good, though!

MSLilah commented:
I'm :+1: barring the merging of those two schemas. I'm fine with punting work around error handling in the callback functions to a future PR dealing with TK-175.

rlinehan commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -13,18 +13,27 @@
+   :service_status_version schema/Int
+   :is_running (schema/enum true false :unknown)
+   :detail_level ServiceStatusDetailLevel
    :status schema/Any})
```
Ergh, so the problem with reusing `StatusCallbackResponse` in this is that this uses underscores (because ultimately it's going to be JSON output), while that has hyphens. We could just use underscores everywhere, I guess? Or just keep them separate.

rlinehan committed:
(TK-209) Make value for :is-running a keyword

Add an `IsRunning` schema that can be reused in multiple places. This schema
changes the value of `:is-running` to always be a keyword: :true, :false, or
:unknown.

Add a TODO for merging the `StatusCallbackResponse` schema into the `Status`
schema, rather than duplicating its two keys. In rder to do this, we need to
make some changes to how we change hyphens to underscores for json output, but
for now leaving that for another time.

rlinehan committed:
(maint) Add comment about pmap exceptions

rlinehan commented:
I agree about punting error handling work until TK-175, since I think that ticket is meant for that, and I think we have a little more discussion as to what the right behavior is (although I do agree that it should be in a try/catch).

I've added a commit that makes a schema for the output of `:is-running`, so that it's always a keyword. Unfortunately, I think before we can merge those two schemas we need to have a better way of doing hyphen -> underscore transformation for json output (and I didn't really want to do that at 5pm). However, I think that at this point doing that (and merging these schemas) should not cause any API changes, so hopefully that's not a blocker?

cprice404 commented:
I think you're right that the API should hopefully not be affected when we come back around and deal with the JSON/underscore/combining schemas madness.  Thanks!

cprice404 commented:
@rlinehan do you think it would be worth starting an e-mail thread between jade and the NC team to ask for their advice / warnings re: underscores and JSON?
